### PR TITLE
Set up tests for admission webhook

### PR DESCRIPTION
**Pull request by JoelSpeed**
This PR adds a lot of boilerplate around testing as well as an initial set of high level tests around the webhooks for the new ControlPlaneMachineSet resource

Some notes on the commits that support these new tests:
- Update the API to include restrictive template objectmeta
  - Because we were using ObjectMeta from metav1, there are a number of fields that aren't actually used by the project. This commit restricts the fields within that struct to only those actually used in the project
- Add komega - Currently WIP for merging into controller-runtime, will remove once merged into controller-runtime
- Add resourcebuilders - This allows us to easily generate resources within test cases, to be expanded as new types are added
- Add cleanup utility - This can be called at the end of each test to cleanup the environment and resources created within the test, without having to track the resources, only their kinds

### CODE REVIEW

**Review by elmiko — COMMENTED** [review]
this looks really nice to me. i especially like the fluent builder style you've implemented for the test case setup stuff. i have an inline question and i'm also curious if the `TODO`s in the tests are just some placeholders for the work you are following up with?

/approve

File: pkg/test/suite_test.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	RunSpecs(t, "Test Suite")
+}
+
+var _ = BeforeSuite(func() {
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+
+	By("bootstrapping test environment")
+	testEnv = &envtest.Environment{
+		CRDDirectoryPaths:     []string{filepath.Join("..", "..", "vendor", "github.com", "openshift", "api", "machine", "v1beta1")},
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply pkg/test/suite_test.go:55]: I will add the v1 group to the test suites when we need v1 in this project, for now, we aren't importing v1 at all, but will when the CPMS ends up in openshift/api

**Comment by JoelSpeed** [comment]
>  have an inline question and i'm also curious if the TODOs in the tests are just some placeholders for the work you are following up

The TODOs are because I don't know what the errors will look like yet, I want to see an error here and have the error string somewhat recognised, eg it could say "label selector is immutable after creation", but as that's not implemented yet, I'll leave it to whoever implements the webhooks to sort

**Comment by JoelSpeed** [comment]
/hold 

Going to update this because the komega code is about to merge into controller-runtime

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-aws-upgrade

This isn't in payload yet so this isn't actually doing anything right now

**Pull request merged**